### PR TITLE
add USB host driver function for imxrt

### PR DESCRIPTION
**Pull request by HarrisonChoi750405**
1) add USB host initialization
2) add USB ehci interface

### CODE REVIEW

File: os/arch/arm/src/imxrt/imxrt_usbhost.c [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+ ***********************************************************************************/
+
+void imxrt_usbhost_vbusdrive(int rhport, bool enable)
+{
+	uinfo("RHPort%d: enable=%d\n", rhport + 1, enable);
+
+	/* The LPC3131 has only a single root hub port */
+
+#if 0
+	if (rhport == 0) {
```
Comment (yashwanth686007) [inline os/arch/arm/src/imxrt/imxrt_usbhost.c]: Please remove the code if not required.

File: os/arch/arm/src/imxrt/imxrt_usbhost.c [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+	if (!g_ehciconn) {
+		IMXLOG("ERROR: imxrt_ehci_initialize failed\n");
+		return -ENODEV;
+	}
+
+	/* Start a thread to handle device connection. */
+	pid = task_create("EHCI Monitor", CONFIG_USBHOST_DEFPRIO, CONFIG_USBHOST_STACKSIZE, (main_t) ehci_waiter, (FAR char *const *)NULL);
+	if (pid < 0) {
+		IMXLOG("ERROR: Failed to create ehci_waiter task: %d\n");
+		return -ENODEV;
```
Comment (yashwanth686007) [inline os/arch/arm/src/imxrt/imxrt_usbhost.c]: Its better to return the appropriate error value.

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 3]
```diff
@@ -0,0 +1,5148 @@
...
+/* Configurable number of Queue Element Transfer Descriptor (qTDs).  The default
+ * is one per root hub plus three from EP0.
+ */
+
+#ifndef CONFIG_IMXRT_EHCI_NQTDS
+#define CONFIG_IMXRT_EHCI_NQTDS (IMXRT_EHCI_NRHPORT + 3)
+#endif
+
+/* Buffers must be aligned to the cache line size */
+#define DCACHE_LINEMASK (ARMV7M_DCACHE_LINESIZE -1)
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_ehci.c]: please provide space before '-'

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 4]
```diff
@@ -0,0 +1,5148 @@
...
+ *
+ * Description:
+ *   Read 16-bit little endian data
+ *
+ ****************************************************************************/
+
+static uint16_t imxrt_read16(const uint8_t *addr)
+{
+#ifdef CONFIG_ENDIAN_BIG
+	return (uint16_t) addr[0] << 8 | (uint16_t) addr[1];
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_ehci.c]: please remove space after typecasting. 
like: (uint16_t)addr[0] << 8 | (uint16_t)addr[1];

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 5]
```diff
@@ -0,0 +1,5148 @@
...
+ *
+ * Description:
+ *   Print the contents of a LPC31 EHCI register
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_IMXRT_EHCI_REGDEBUG
+static void imxrt_printreg(volatile uint32_t *regaddr, uint32_t regval, bool iswrite)
+{
+	uinfo("%08x%s%08x\n", (uintptr_t) regaddr, iswrite ? "<-" : "->", regval);
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_ehci.c]: same as above. and please check at all places

File: os/arch/arm/src/imxrt/imxrt_clockconfig.c [hunk 6]
```diff
@@ -92,6 +109,61 @@ const clock_usb_pll_config_t usb1PllConfig_BOARD_BootClockRUN = {
 	.src = 0,         /* Bypass clock source, 0 - OSC 24M, 1 - CLK1_P and CLK1_N */
 };
 
+const usb_phy_config_struct_t PhyConfig = {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_clockconfig.c]: It is clock config, could you please explain why usb feature added here?

Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_clockconfig.c]: No specific reason.
From boot sequence in iMXRT, usb phy initialization need to be done after completing clock configuration

File: os/arch/arm/src/imxrt/imxrt_clockconfig.c [hunk 7]
```diff
@@ -92,6 +109,61 @@ const clock_usb_pll_config_t usb1PllConfig_BOARD_BootClockRUN = {
...
+ *          for example: if the external input clock is 16M, the parameter freq should be 16000000.
+ *
+ * Returned Value:
+ *   Returns kStatus_USB_Success      cancel successfully.
+ *   Returns kStatus_USB_Error        the freq value is incorrect.
+ *
+ ****************************************************************************/
+uint32_t imxrt_usb_ehciphyinit(uint32_t freq, usb_phy_config_struct_t *phyConfig)
+{
+#if ((defined FSL_FEATURE_SOC_USBPHY_COUNT) && (FSL_FEATURE_SOC_USBPHY_COUNT > 0U))
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_clockconfig.c]: I cannot understand your point.
Please let me know in detail.

File: os/arch/arm/src/imxrt/imxrt_usbhost.c (hunk 1 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_usbhost.c]: I want to keep this code for reference.
I will add another definition for it
Please check it

File: os/arch/arm/src/imxrt/imxrt_ehci.c (hunk 5 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_ehci.c]: fixed it

File: os/arch/arm/src/imxrt/imxrt_ehci.c (hunk 4 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_ehci.c]: fixed it

File: os/arch/arm/src/imxrt/imxrt_ehci.c (hunk 3 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_ehci.c]: fixed it

File: os/arch/arm/src/imxrt/imxrt_usbhost.c (hunk 2 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_usbhost.c]: fixed it

File: build/configs/imxrt1050-evk/nxp_demo/defconfig [hunk 8]
```diff
@@ -296,7 +304,16 @@ CONFIG_IMXRT_AUTOMOUNT_USERFS=y
 CONFIG_IMXRT_AUTOMOUNT_USERFS_DEVNAME="/dev/smart0p8"
 CONFIG_IMXRT_AUTOMOUNT_USERFS_MOUNTPOINT="/mnt"
 # CONFIG_IMXRT_AUTOMOUNT_SSSRW is not set
-# CONFIG_IMXRT_AUTOMOUNT_ROMFS is not set
+CONFIG_IMXRT_AUTOMOUNT_ROMFS=y
+CONFIG_IMXRT_AUTOMOUNT_ROMFS_DEVNAME="/dev/mtdblock9"
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply build/configs/imxrt1050-evk/nxp_demo/defconfig]: fixed it

File: build/configs/imxrt1020-evk/nxp_demo/defconfig [hunk 9]
```diff
@@ -300,7 +302,9 @@ CONFIG_IMXRT_AUTOMOUNT_USERFS=y
 CONFIG_IMXRT_AUTOMOUNT_USERFS_DEVNAME="/dev/smart0p1"
 CONFIG_IMXRT_AUTOMOUNT_USERFS_MOUNTPOINT="/mnt"
 # CONFIG_IMXRT_AUTOMOUNT_SSSRW is not set
-# CONFIG_IMXRT_AUTOMOUNT_ROMFS is not set
+CONFIG_IMXRT_AUTOMOUNT_ROMFS=y
+CONFIG_IMXRT_AUTOMOUNT_ROMFS_DEVNAME="/dev/mtdblock9"
+CONFIG_IMXRT_AUTOMOUNT_ROMFS_MOUNTPOINT="/rom"
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply build/configs/imxrt1020-evk/nxp_demo/defconfig]: fixed it

**Comment by seinfra** [comment]
Target : [9e5f948ab7895b6410f3c5ebbe86201a644c99a2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [9e5f948ab7895b6410f3c5ebbe86201a644c99a2] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_ehci.c:3204: ERROR: [SPC_M_KWD] space required before the open brace '{'
os/arch/arm/src/imxrt/imxrt_ehci.c:3204: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/imxrt/imxrt_ehci.c:3412: ERROR: [SPC_M_KWD] space required before the open brace '{'
os/arch/arm/src/imxrt/imxrt_ehci.c:3412: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/imxrt/imxrt_ehci.c:3546: ERROR: [SPC_M_KWD] space required before the open brace '{'
os/arch/arm/src/imxrt/imxrt_ehci.c:3546: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [74c06b7c395d60c9d481bcfbcd5d8e6459c8c4db] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 10]
```diff
@@ -0,0 +1,5148 @@
...
+ * Name: imxrt_read32
+ *
+ * Description:
+ *   Read 32-bit little endian data
+ *
+ ****************************************************************************/
+static inline uint32_t imxrt_read32(const uint8_t *addr)
+{
+#ifdef CONFIG_ENDIAN_BIG
+	return (uint32_t) imxrt_read16(&addr[0]) << 16 | (uint32_t) imxrt_read16(&addr[2]);
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_ehci.c]: Please remove space after typecasting.

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 11]
```diff
@@ -0,0 +1,5148 @@
...
+ *
+ * Description:
+ *   Swap bytes on a 32-bit value
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ENDIAN_BIG
+static uint32_t imxrt_swap32(uint32_t value)
+{
+	return (uint32_t) imxrt_swap16((uint16_t)((value >> 16) & 0xffff)) | (uint32_t) imxrt_swap16((uint16_t)(value & 0xffff)) << 16;
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_ehci.c]: same as above.

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 12]
```diff
@@ -0,0 +1,5148 @@
...
+	 */
+
+	qh->hw.hlp = qhead->hw.hlp;
+	imxrt_qh_flush(qh);
+
+	/* Then set the new QH as the first QH in the asychronous queue and flush the
+	 * modified head to RAM.
+	 */
+
+	physaddr = (uintptr_t) imxrt_physramaddr((uintptr_t) qh);
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_ehci.c]: same as above, Please check it for entire file.

**Comment by seinfra** [comment]
Target : [74c06b7c395d60c9d481bcfbcd5d8e6459c8c4db] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_ehci.c:3204: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [74c06b7c395d60c9d481bcfbcd5d8e6459c8c4db] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [74c06b7c395d60c9d481bcfbcd5d8e6459c8c4db] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [802701dc171ea3093921a61fc8709231fba1ed81] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [802701dc171ea3093921a61fc8709231fba1ed81] - Code Rule Check OK.

**Pull request merged**